Scala.js support

guymers opened the pull request:
A series of patches that enable Scala.js support plus a simple backend implementation based on the fetch api. I have not yet had the chance to test with a real project.

Summary of commits:

- Extract MonadAsyncError implementations
Moves implementations into their own projects which allows sharing between the JVM and JS when applicable

- Move backend tests into their projects
Moves the test http server into its own project which allows JS to access it

- Change HttpTest to AsyncFreeSpec
JS cannot block so all tests must be async

- Remove ResponseAsFile
Saving a response to a file does not make sense in a browser context plus this functionality can be easily implementation by using a streaming backend

- Remove InputStreamBody
It is better to use a streaming backend as it is asynchronous 

- Scala.js support
Cross builds core, monix, cats and circe for Scala.js. Cookies are JVM only due to their dependencies on Java only classes (the fetch api cannot access the cookie headers plus will pass them in the request). Files are handled via a wrapper class which has implicit conversions from java.io.File, java.nio.files.Path and JS web File. IDN support is provided by the JS URL class which Scala.js does not seem to be able to find on nodejs 8 so at least 10 is required to run the tests.

- Fetch backend for Scala.js
A simple non streaming backend using the fetch api. Timeouts are implemented via AbortController which was only added to Chrome in the latest release (66). Running the tests therefore requires you have at least Chrome 66 plus chromedriver on the path. Will look at adding a streaming implementation in the future.

I had to disable the akka client test from running in CI due to https://github.com/akka/akka-http/issues/1930 causing the build to hang. It is still present in the root aggregate project.

adamw commented:
Thanks a lot for the changes! There's quite a lot of things in there, so just some high-level comments for now:

* I'd definitely like to keep `File` and `InputStream` support as-is. They are just very common on the JVM and very useful. 
* as for the JS backend implementation, I saw that it relies on the fetch APIs which are marked experimental. I'm not very up-to-date with the JS ecosystem, but will that actually work on browsers/nodejs? Is there maybe another way to implement the backend using a more compatible API?
* so far I've kept the `MonadError` implementations local to the backends and copy-pasted when needed - it's not a lot of code but simplifies the build, which already has a lot of pretty small modules
* do you think there would a way to split the PR into smaller ones with smaller surface areas?

guymers commented:
- Okay. Having `File` support requires the library to add extra code to handle save options such as overwriting. Async streaming implementations will need to convert to the blocking `InputStream`. I think its better for the caller to be required to do that complex operation. I also thought that Scala.js did not support `InputStream` but I obviously didn't look closely.
- `fetch` is supported in all mainstream browsers and has been for some time. There is a nodejs implementation available on npm. The browser could be implemented via `XMLHttpRequest` but `fetch` has proper streaming support (behind a flag on Firefox). 
- There were only two locations before but I would like to make a Monix `fetch` implementation in the future. With the tests being located with each backend it also avoids having to duplicate some test code as well.
- It is currently in 7 commits, I think they each focus on a single area pretty well? Do you want me to make a PR with the first one to review on its own?

adamw commented:
* I don't really know if `File` and `InputStream` are supported in Scala.JS - maybe they aren't. Will the code compile at all for scalajs if there are methods accepting these types? I think they are common enough in the JVM that they need to be supported, even if it might mean a potential run-time exception when trying to use them from js
* ok, thanks, that's good enough. Though according to https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API IE is not supported? ;)
* as for extracting monad impls to separate projects maybe that's a good thing to do but as the PR is already very big and reorganizes a lot of the project I was thinking of ways to slim it down - that might be one possibility. If you could create more focused PRs that would definitely be great, if not I think we can try to work with this one

guymers commented:
I have reverted the removal of `InputStream` and `File` and added support in the `fetch` backend.

Yeah no IE, don't think that is really a problem these days.

I have created another PR with the first patch: https://github.com/softwaremill/sttp/pull/93

adamw commented:
Now that #93 is in I suppose this one needs an update :)

guymers commented:
2nd patch PR https://github.com/softwaremill/sttp/pull/94

This PR can be the 4th one ;)
